src/transformers/generation/utils.py

```diff
@@ -4915,3 +5242,76 @@ def _concat(data):
 
     # Return a new object of the inferred class with the concatenated attributes
     return model_output_cls(**concatenated_data)
+
+
+def _dola_select_contrast(
+    candidate_premature_layers: List[int],
+    candidate_premature_logits: Dict[int, torch.FloatTensor],
+    final_logits: torch.FloatTensor,
+) -> torch.FloatTensor:
+    if len(candidate_premature_layers) == 1:
+        base_logits = candidate_premature_logits[candidate_premature_layers[0]]
+        final_logits, base_logits = _relative_top_filter(final_logits, base_logits)
+        logits = final_logits - base_logits
+    else:
+        # 1. Stacking all premature_layers into a new dimension
+        stacked_premature_layers = torch.stack(
+            [candidate_premature_logits[i] for i in candidate_premature_layers], dim=0
+        )
+
+        # 2. Calculate the softmax values for mature_layer and all premature_layers
+        softmax_mature_layer = F.softmax(final_logits, dim=-1)  # shape: (batch_size, vocab_size)
+        softmax_premature_layers = F.softmax(
+            stacked_premature_layers, dim=-1
+        )  # shape: (num_premature_layers, batch_size, vocab_size)
+
+        # 3. Calculate M, the average distribution
+        M = 0.5 * (
```

Comment:
As a rule, no single letter vars should be used - let's use something more descriptive e.g. `avg_dist`

Reply:
Changed it to `avg_dist`!